### PR TITLE
mgr/dashboard: stay on active tab (accessible via URL)

### PR DESCRIPTION
**Pull request by votdev**
This PR refactors the /monitoring page by introducing a new component which is responsible for rendering the tabs. All tabs have a dedicated route/URL now, no fragments are used anymore, thus they will act like the iSCI and RBD pages.

A helper class, used by the `Active Alerts`, `Alerts` and `Silences` page, has been introduced to prevent duplicate code. This class itself derives from `ListWithDetails` to prevent the need to use a Typescript Mixin framework, e.g. [ts-mixer](https://www.npmjs.com/package/ts-mixer). There is no need to add another dependency.

screenshot

Fixes: https://tracker.ceph.com/issues/45375

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by LenzGr** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/shared/classes/tab-helper.class.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import { TabDirective, TabsetComponent } from 'ngx-bootstrap/tabs';
+
+export class TabHelper implements OnInit {
+  @ViewChild(TabsetComponent, { static: true })
+  tabset: TabsetComponent;
+
+  constructor(protected route: ActivatedRoute, protected router: Router) {}
+
+  ngOnInit() {
+    const fragment = this.route.snapshot.fragment;
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/classes/tab-helper.class.ts:13]: I assume that fragment should actually be passed to the class (?)

Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/classes/tab-helper.class.ts:13]: `fragment`is a member variable of `ActivatedRouteSnapshot`, but it must be set in a `this.router.navigate(['/monitoring'], { fragment: 'silences' });` command.

**Comment by votdev** [comment]
> Just a suggestion (a change in the current focus: using `tab` inputs to activate them instead the output methods, and rely on RouterLink to do the rest).
> 
> That's how [ng-bootrstap](https://ng-bootstrap.github.io/#/components/nav/overview#routing) and [Angular Material](https://material.angular.io/components/tabs/overview#tabs-and-navigation) make it with RouterLink...
> 
> For example, what about this?:
> 
> ```
> <tabset routerLinkActive #rla="routerLinkActive">
>   <tab id="active-alerts" routerLink="." [fragment]="attr.id" [active]="rla.isActive">
>   </tab>
>   <tab id="all-alerts" routerLink="."  [fragment]="attr.id" [active]="rla.isActive">
>   </tab>
> </tabset>
> ```
> 
> (Not a 100% sure everything will work there: it might need some tuning).
> 
> Pros I see from that approach:
> 
>     * No need to explicitly use inheritance and derived components, nor messing with routing services within components. I'm a fan of OOP, but for related concerns (I think navigation is an unrelated concern here).
> 
>     * With an optional custom directive (dealing with the routerlink and fragment mapping to ids), you can further minimize the code changes and automatically apply this behaviour to all tabsets/tabs.
> 
>     * You keep the Router Links exposed in the HTML, as with regular `<a>` elements.

That does not work because we are using `ngx-bootstrap` where `tabset` does not support routing. Need to find a workaround.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/classes/tab-helper.class.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      // Ensure fragment is not removed from URL, so page can always be
+      // reloaded with the same tab open.
+      this.router.navigate([], { fragment });
+    }
+  }
+
+  setFragment(element: TabDirective) {
+    this.router.navigate([], { fragment: element.id });
+  }
+}
```
[the remark this answers is not in the document]
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/classes/tab-helper.class.ts]: Fixed by introducing two new directives.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-activate-by-fragment.directive.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * fragment to the current route, e.g. `/monitoring#silences`.
+ */
+@Directive({
+  selector: '[cdTabActivateByFragment]'
+})
+export class TabActivateByFragmentDirective implements OnInit {
+  constructor(@Optional() @Host() private tab: TabDirective, private router: Router) {}
+
+  @Input()
+  fragment: string;
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-activate-by-fragment.directive.ts:20]: As mentioned: given we don't really care too much the actual "text" for the fragment, why not simply reuse the same we have for "heading"?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-activate-by-fragment.directive.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  ngOnInit() {
+    const urlTree: UrlTree = this.router.parseUrl(this.router.url);
+    if (urlTree.fragment === this.fragment) {
+      _.set(this.tab, 'active', true);
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-activate-by-fragment.directive.ts:25]: Shouldn't we set also the `id`? In this it works great because the tabs are placed in the top of the page, but if not, the browser won't scroll to the "#" anchor.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-activate-by-fragment.directive.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import { TabDirective } from 'ngx-bootstrap/tabs';
+
+/**
+ * This directive will activate the tab if the route fragment is
+ * identical to its identifier. It will also navigate to the current
+ * route when the tab is selected by adding the tab identifier as
+ * fragment to the current route, e.g. `/monitoring#silences`.
+ */
+@Directive({
+  selector: '[cdTabActivateByFragment]'
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-activate-by-fragment.directive.ts:14]: I think we could even 'automagically' apply this to all `<tabs>` by simply reusing the same selector pattern that 'TabDirective uses (Angular allows to apply multiple directives to the same selector):
 
```suggestion
  selector: 'tab:[tab]'
```

I was also going to suggest using inheritance here (instead of DI, so you don't need to redefine `@Input() fragment`), but given we don't control ngx-bootstrap, it think this is a safer approach.

**Comment by tspmelo** [comment]
Sorry for the late review.

We currently have a different approach for this problem in the block images pages and I think I like it better. 

Few things to consider:
- Only the active tab should be loaded, to reduce memory consumption and requests made
  - RBD implementation this happens automatically, since only 1 component is loaded
  - Should be easy to implement in this solution
- RBD has a cleaner code in the "index" component, like  `rbd-tabs` and  in this case would be `monitoring-list`
- RBD doesn't have 2 fragments in the URL
  - although this seems to work, it looks strange to me
- `rbd-tabs.component.html` could be improved further by refactoring the code used in [active] and (selectTab) by declaring the url only once.
- In RBD we can find out all the possible URL of ceph dashboard by going through routing declarations, currently don't think it matter, but could be used in a future for something like a sitemap
- In RBD changing tab also updates the breadcrumb

Independently of which solution we decide on, I think we should refactor all components similar to these ones to use the same.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-activate-by-fragment.directive.ts (hunk 5 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-activate-by-fragment.directive.ts:14]: Will this also apply to tabs inside details? don't think we should do this.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/monitoring-list/monitoring-list.component.html [hunk 6]
```diff
@@ -1,7 +1,7 @@
-<tabset #tabs>
-  <tab id="active-alerts"
+<tabset>
+  <tab cdTabActivateByFragment
+       fragment="active-alerts"
        heading="Active Alerts"
```
[the remark this answers is not in the document]
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/monitoring-list/monitoring-list.component.html:4]: > Sorry for the late review.
> 
> We currently have a different approach for this problem in the block images pages and I think I like it better.
> 
> Few things to consider:
> 
>     * Only the active tab should be loaded, to reduce memory consumption and requests made
>       
>       * RBD implementation this happens automatically, since only 1 component is loaded
>       * Should be easy to implement in this solution
> 
>     * RBD has a cleaner code in the "index" component, like  `rbd-tabs` and  in this case would be `monitoring-list`
> 
>     * RBD doesn't have 2 fragments in the URL
>       
>       * although this seems to work, it looks strange to me
> 
>     * `rbd-tabs.component.html` could be improved further by refactoring the code used in [active] and (selectTab) by declaring the url only once.
> 
>     * In RBD we can find out all the possible URL of ceph dashboard by going through routing declarations, currently don't think it matter, but could be used in a future for something like a sitemap
> 
>     * In RBD changing tab also updates the breadcrumb
> 
> 
> Independently of which solution we decide on, I think we should refactor all components similar to these ones to use the same.

I will refactor the monitoring-list like it is done in rbd using a `monitoring-tabs` component. I think this will be the better approach.

**Comment by votdev** [comment]
> Sorry for the late review.
> 
> We currently have a different approach for this problem in the block images pages and I think I like it better.
> 
> Few things to consider:
> 
>     * Only the active tab should be loaded, to reduce memory consumption and requests made
>       
>       * RBD implementation this happens automatically, since only 1 component is loaded
>       * Should be easy to implement in this solution
> 
>     * RBD has a cleaner code in the "index" component, like  `rbd-tabs` and  in this case would be `monitoring-list`
> 
>     * RBD doesn't have 2 fragments in the URL
>       
>       * although this seems to work, it looks strange to me
> 
>     * `rbd-tabs.component.html` could be improved further by refactoring the code used in [active] and (selectTab) by declaring the url only once.
> 
>     * In RBD we can find out all the possible URL of ceph dashboard by going through routing declarations, currently don't think it matter, but could be used in a future for something like a sitemap
> 
>     * In RBD changing tab also updates the breadcrumb
> 
> 
> Independently of which solution we decide on, I think we should refactor all components similar to these ones to use the same.

I will refactor the `monitoring-list` by using a `monitoring-tabs` component like it is done for RBD. I think this will be the better approach.

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
The PR must be refactored/adapted after we changed to the new ng-bootstrap library.

**Comment by tspmelo** [comment]
typo on the commit message: `s/iSCI/iSCSI`

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/silence-list/silence-list.component.ts [hunk 7]
```diff
@@ -52,14 +54,15 @@ export class SilenceListComponent extends ListWithDetails {
     private authStorageService: AuthStorageService,
     private i18n: I18n,
     private cdDatePipe: CdDatePipe,
-    private prometheusService: PrometheusService,
-    private modalService: ModalService,
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/silence-list/silence-list.component.ts:56]: Please revert this change.

**Comment by votdev** [comment]
jenkins test make check
